### PR TITLE
Java: CWE-552 Add sources and sinks to to detect unsafe getResource calls in Java EE applications

### PR DESCRIPTION
**Pull request by luchua-bc**
Directly incorporating user input into the getResource call on the server side without proper validation of the input can allow any web application resource such as configuration files and source code to be disclosed.

It is very common that Java EE applications load requested resources and return file contents to clients. Constructing a server-side URI path with user input could allow an attacker to download application binaries
(including application classes or jar files) or view arbitrary files within protected directories.
 
This query detects unsafe usage of `getResource` or `getResourceAsStream` from both servlet context and Java Server Faces context. It utilizes the path sanitizer library to reduce FPs:
- Path traversal check
- Path encoding check
- Check of path normalization using the java.nio.file.Path package

Please consider to merge the PR. Thanks.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/frameworks/Servlets.qll [hunk 1]
```diff
@@ -377,3 +377,26 @@ class RequestDispatchMethod extends Method {
     this.hasName(["forward", "include"])
   }
 }
+
+/**
+ * The interface `javax.servlet.ServletContext`.
+ */
+library class ServletContext extends RefType {
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/frameworks/Servlets.qll:384]: Don't use `library`, it's deprecated.

```suggestion
class ServletContext extends RefType {
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+InputStream in = request.getServletContext().getResourceAsStream(requestPath);
+in = getClass().getClassLoader().getResourceAsStream(requestPath);
+
+// GOOD: check for a trusted prefix, ensuring path traversal is not used to erase that prefix:
+// (alternatively use `Path.normalize` instead of checking for `..`)
+if (!requestPath.contains("..") && requestPath.startsWith("/trusted")) {
+	InputStream in = request.getServletContext().getResourceAsStream(requestPath);
+}
+
+Path path = Paths.get(requestUrl).normalize().toRealPath();
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java:15]: Isn't this lacking the `startsWith` check (or similar) to be considered secure?

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// BAD: no URI validation
+URL url = servletContext.getResource(requestUrl);
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java:2]: Use `request.getServletContext()` throughout the example for consistency.

```suggestion
URL url = request.getServletContext().getResource(requestUrl);
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+in = getClass().getClassLoader().getResourceAsStream(requestPath);
+
+// GOOD: check for a trusted prefix, ensuring path traversal is not used to erase that prefix:
+// (alternatively use `Path.normalize` instead of checking for `..`)
+if (!requestPath.contains("..") && requestPath.startsWith("/trusted")) {
+	InputStream in = request.getServletContext().getResourceAsStream(requestPath);
+}
+
+Path path = Paths.get(requestUrl).normalize().toRealPath();
+URL url = sc.getResource(path.toString());
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java:16]:
```suggestion
URL url = request.getServletContext().getResource(path.toString());
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp [hunk 5]
```diff
@@ -47,5 +54,8 @@ attacks. It also shows how to remedy the problem by validating the user input.
 <li>Micro Focus:
   <a href="https://vulncat.fortify.com/en/detail?id=desc.dataflow.java.file_disclosure_j2ee">File Disclosure: J2EE</a>
 </li>
+<li>
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp:57]:
```suggestion
<li>CVE-2015-5174:
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp [hunk 6]
```diff
@@ -47,5 +54,8 @@ attacks. It also shows how to remedy the problem by validating the user input.
...
+  <a href="https://vuldb.com/?id.81084">Apache Tomcat 6.0/7.0/8.0/9.0 Servletcontext Getresource/getresourceasstream/getresourcepaths Path Traversal</a>
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp:58]:
```suggestion
  <a href="https://vuldb.com/?id.81084">Apache Tomcat 6.0/7.0/8.0/9.0 ServletContext getResource/getResourceAsStream/getResourcePaths Path Traversal</a>
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 7]
```diff
@@ -18,6 +19,84 @@ private class RequestDispatcherSink extends UnsafeUrlForwardSink {
   }
 }
 
+/** The `getResource` method of `Class`. */
+class GetClassResourceMethod extends Method {
+  GetClassResourceMethod() {
+    this.getSourceDeclaration().getDeclaringType().hasQualifiedName("java.lang", "Class") and
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:25]:
```suggestion
    this.getDeclaringType() instanceof TypeClass and
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 8]
```diff
@@ -18,6 +19,84 @@ private class RequestDispatcherSink extends UnsafeUrlForwardSink {
...
+    this.hasName("getResource")
+  }
+}
+
+/** The `getResourceAsStream` method of `Class`. */
+class GetClassResourceAsStreamMethod extends Method {
+  GetClassResourceAsStreamMethod() {
+    this.getSourceDeclaration().getDeclaringType().hasQualifiedName("java.lang", "Class") and
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:33]:
```suggestion
    this.getDeclaringType() instanceof TypeClass and
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 9]
```diff
@@ -18,6 +19,84 @@ private class RequestDispatcherSink extends UnsafeUrlForwardSink {
...
+  GetClassLoaderResourceMethod() {
+    this.getDeclaringType().hasQualifiedName("java.lang", "ClassLoader") and
+    this.hasName("getResource")
+  }
+}
+
+/** The `getResourceAsStream` method of `ClassLoader`. */
+class GetClassLoaderResourceAsStreamMethod extends Method {
+  GetClassLoaderResourceAsStreamMethod() {
+    this.getDeclaringType().hasQualifiedName("java.lang", "ClassLoader") and
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:49]:
```suggestion
    this.getDeclaringType() instanceof ClassLoaderClass and
```

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 10]
```diff
@@ -18,6 +19,84 @@ private class RequestDispatcherSink extends UnsafeUrlForwardSink {
...
+  }
+}
+
+/** The JBoss class `VirtualFile`. */
+class VirtualFile extends RefType {
+  VirtualFile() { this.hasQualifiedName("org.jboss.vfs", "VirtualFile") }
+}
+
+/** The JBoss method `getChild` of `FileResourceManager`. */
+class GetVirtualFileMethod extends Method {
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll:75]: This name seems like it's lacking a reference to `getChild`. Maybe it should be something like `GetVirtualFileChildMethod` (following the pattern of the other names above)?

File: java/ql/src/experimental/semmle/code/java/frameworks/Jsf.qll [hunk 11]
```diff
@@ -0,0 +1,34 @@
+/**
+ * Provides classes and predicates for working with the Java Server Faces (JSF).
+ */
+
+import semmle.code.java.Type
+
+/**
+ * The JSF class `FacesContext` for processing HTTP requests.
```
Comment (atorralba) [inline java/ql/src/experimental/semmle/code/java/frameworks/Jsf.qll:8]:
```suggestion
 * The JSF class `ExternalContext` for processing HTTP requests.
```

**Review by atorralba — COMMENTED** [review]
Hey @luchua-bc, thank you for your contribution. I added a few inline comments.

**Comment by luchua-bc** [comment]
Thanks @atorralba for reviewing this PR and your help with merging with main. I've made all requested changes. Please review again.

Thanks,
@luchua-bc

File: java/ql/lib/semmle/code/java/frameworks/Servlets.qll (hunk 1 above)
Comment (luchua-bc) [reply java/ql/lib/semmle/code/java/frameworks/Servlets.qll]: I've made requested change. Please also note the class `ServletContext` itself is not deprecated, which is still widely and actively used in Java EE applications as a main method to interact with the servlet container, although several methods related to `getServlet` and `log` were deprecated. Please let me know if we want to keep the `library` modifier. Thanks.

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java (hunk 2 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java:15]: Yes, it lacks the `startsWith` check and I've added it in the sample usage.

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java (hunk 3 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeResourceGet.java]: Changed for consistency

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp (hunk 5 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp]: Added

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp (hunk 6 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qhelp]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 7 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll [hunk 12]
```diff
@@ -18,6 +19,84 @@ private class RequestDispatcherSink extends UnsafeUrlForwardSink {
...
+  GetClassResourceAsStreamMethod() {
+    this.getSourceDeclaration().getDeclaringType().hasQualifiedName("java.lang", "Class") and
+    this.hasName("getResourceAsStream")
+  }
+}
+
+/** The `getResource` method of `ClassLoader`. */
+class GetClassLoaderResourceMethod extends Method {
+  GetClassLoaderResourceMethod() {
+    this.getDeclaringType().hasQualifiedName("java.lang", "ClassLoader") and
```
[the remark this answers is not in the document]
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 9 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 8 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll (hunk 10 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll]: Changed the method name to `GetVirtualFileChildMethod`

File: java/ql/src/experimental/semmle/code/java/frameworks/Jsf.qll (hunk 11 above)
Comment (luchua-bc) [reply java/ql/src/experimental/semmle/code/java/frameworks/Jsf.qll]: Thanks for correcting the typo. I've made the change.

File: java/ql/lib/semmle/code/java/frameworks/Servlets.qll (hunk 1 above)
Comment (atorralba) [reply java/ql/lib/semmle/code/java/frameworks/Servlets.qll]: What I meant is that the `library` annotation is deprecated in CodeQL. The `ServletContext` can still be used normally.

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql [hunk 13]
```diff
@@ -43,6 +43,20 @@
   override DataFlow::FlowFeature getAFeature() {
     result instanceof DataFlow::FeatureHasSourceCallContext
   }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ) {
+    exists(MethodAccess ma |
+      (
+        ma.getMethod() instanceof GetServletResourceMethod or
+        ma.getMethod() instanceof GetFacesResourceMethod or
```
[the remark this answers is not in the document]
Comment (atorralba) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:52]: It seems that `UnsafeUrlForward.ql` needs an import of `Jsf.qll`.

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql [hunk 14]
```diff
@@ -43,6 +44,20 @@ class UnsafeUrlForwardFlowConfig extends TaintTracking::Configuration {
   override DataFlow::FlowFeature getAFeature() {
     result instanceof DataFlow::FeatureHasSourceCallContext
   }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ) {
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:48]: These taint steps should probably have tests.

**Review by atorralba — COMMENTED** [review]
A couple more comments.

File: java/ql/lib/semmle/code/java/frameworks/Servlets.qll (hunk 1 above)
Comment (luchua-bc) [reply java/ql/lib/semmle/code/java/frameworks/Servlets.qll]: Thanks @atorralba for the clarification.

File: java/ql/src/experimental/semmle/code/java/frameworks/Jsf.qll [hunk 15]
```diff
@@ -0,0 +1,34 @@
+/**
+ * Provides classes and predicates for working with the Java Server Faces (JSF).
+ */
+
+import semmle.code.java.Type
```
[the remark this answers is not in the document]
Comment (luchua-bc) [reply java/ql/src/experimental/semmle/code/java/frameworks/Jsf.qll]: Changed

File: java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql (hunk 14 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql:48]: I've added missing test cases for taint steps of class loader, JSF, and JBoss. Please review. Thanks.

**Comment by smowton** [comment]
The PR description reads like it's purely adding more models of sources and sinks, but the test expectation changes remove a result -- is that expected? Are you fixing an FP here at the same time?

**Comment by luchua-bc** [comment]
Thanks @smowton for reviewing this PR. 

I think you are talking about the following test result:
```
| UnsafeServletRequestDispatch.java:110:53:110:76 | toString(...) | UnsafeServletRequestDispatch.java:106:17:106:44 | getParameter(...) : String | UnsafeServletRequestDispatch.java:110:53:110:76 | toString(...) | Potentially untrusted URL forward due to $@. | UnsafeServletRequestDispatch.java:106:17:106:44 | getParameter(...) | user-provided value |
```

Its relevant test case is no longer listed as a `BAD` case: https://github.com/github/codeql/blob/811a2c005304dfa61220b53142cc127ba28a6538/java/ql/test/experimental/query-tests/security/CWE-552/UnsafeServletRequestDispatch.java#L103-L112

Changes in this PR seems to be irrelevant. I'm curious whether a change to a shared library somewhere caused the behaviour change. I will further investigate then report back.

**Comment by luchua-bc** [comment]
Somehow the following taint step was removed during code merge. I've put it back and the issue is resolved. 
```
"java.nio.file;Path;true;toString;;;Argument[-1];ReturnValue;taint"
```

And thanks for pointing out this issue.

@luchua-bc

**Comment by luchua-bc** [comment]
Somehow the following taint step was removed during code merge. I've put it back and the issue is resolved. 
```
"java.nio.file;Path;true;toString;;;Argument[-1];ReturnValue;taint"
```

And thanks for pointing out this issue.

@luchua-bc

**Review by smowton — APPROVED** [review]
LGTM; started tests running.

**Pull request merged**